### PR TITLE
Show a cart abandonment nudge (notice) to users

### PR DESCRIPTION
**Pull request by travisw**
As my first journey into Calypso development I'm trying to show a notice to a user who has an item in their cart older than a certain amount of time. A cart abandonment nudge.
This will be A/B tested on 50% of users.

What this PR does is:
* Start the cart polling which calls the me/shopping-cart/ endpoint to get the cart data (normally polled in other areas of Calypso).
* Check the timestamp (a recently added property) of each item in the cart to see if it is older than 10 minutes.
* If a stale item is found, and the user hasn't already seen the notice recently then show a notice. The notice should click through to the checkout page where they can complete their purchase.

This adds the `lastTimeShown` property to the state tree at `state.notices` which should keep track of the last time any global notice was shown (in milliseconds).

Any feedback is appreciated. :)

Do you see any issues with the shopping cart polling now happening everywhere under My Sites rather than only in a few sections?

**Test Plan**

* From the browser console add yourself to the test variation.
```
localStorage.setItem( 'ABTests', '{"showCartAbandonmentNotice_20170630":"showNotice"}' );
```
* With a test user and site, add any plan to your cart.
* Wait 10+ minutes.
* Click "My Site" to navigate to the stats page.
* As soon as the cart data is synced again (~30 sec.), you should see a notice in the top right of the page similar to this.screen shot 2017-06-30 at 6 49 28 pm
* Click the "Complete your purchase" button in the notice to make sure it takes you back to the checkout page.
* If you have Redux DevTools installed on your browser check to make sure the timestamp in `notices.lastTimeShown.stale-cart-item-notice` has updated.screen shot 2017-07-05 at 4 24 15 pm

### CODE REVIEW

File: client/lib/cart/store/cart-synchronizer.js [hunk 1]
```diff
@@ -244,4 +247,21 @@ CartSynchronizer.prototype.hasPendingServerUpdates = function() {
 	);
 };
 
+CartSynchronizer.prototype._showNoticeForStaleCartItems = function() {
+	const cartItems = require( 'lib/cart-values' ).cartItems;
+
+	if ( cartItems.hasStaleItem( this._latestValue ) ) {
+		debug( 'Firing notice!' );
+		infoNotice(
```
Comment (travisw) [inline client/lib/cart/store/cart-synchronizer.js]: I see now that calling this in this way makes no sense. :)

**Comment by gwwar** [comment]
> As my first journey into Calypso development I'm trying to show a notice to a user who has an item in their cart older than a certain amount of time.

If we don't have one already I think it might make more sense to do this as an email job. Since we can also ping folks who might rarely visit.

**Comment by travisw** [comment]
Thank @gwwar 
> If we don't have one already I think it might make more sense to do this as an email job. Since we can also ping folks who might rarely visit.

I agree that email does make a lot of sense for this and that email does actually already exist.
Here's some background on the different channels being explored.
https://marketingscratchpad.wordpress.com/2017/04/03/cart-abandoner-remarketing-channels/

File: client/my-sites/current-site/index.jsx [hunk 2]
```diff
@@ -67,6 +73,36 @@ class CurrentSite extends Component {
...
+
+		// Show a notice if there are stale itmes in the cart and it hasn't been shown in the last 10
+		if ( cartItems.hasStaleItem( CartStore.get() ) && this.props.staleCartNoticeLastTimeShown < Date.now() - ( 10 * 60 * 1000 ) ) {
+			this.props.infoNotice(
+				this.props.translate( 'Your site deserves a boost!' ),
+				{
+					id: staleCartItemNoticeId,
+					isPersistent: true,
+					duration: 10000,
+					button: this.props.translate( 'Complete your purchase' ),
```
Comment (a8ci18n) [inline client/my-sites/current-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Complete your paypal purchase' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/global-notice-for-cart-abandonment)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/current-site/index.jsx [hunk 3]
```diff
@@ -67,6 +74,36 @@ class CurrentSite extends Component {
...
+		// Show a notice if there are stale itmes in the cart and it hasn't been shown in the last 10 (cart abandonment)
+		if ( cartItems.hasStaleItem( CartStore.get() ) && this.props.staleCartItemNoticeLastTimeShown < Date.now() - ( 10 * 60 * 1000 ) ) {
+			if ( abtest( 'showCartAbandonmentNotice' ) === 'showNotice' ) {
+				this.props.infoNotice(
+					this.props.translate( 'Your site deserves a boost!' ),
+					{
+						id: staleCartItemNoticeId,
+						isPersistent: false,
+						duration: 10000,
+						button: this.props.translate( 'Complete your purchase' ),
```
Comment (a8ci18n) [inline client/my-sites/current-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Complete your paypal purchase' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/global-notice-for-cart-abandonment)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by markryall** [comment]
One little testing tip: redux `state` and `dispatch` are both attached to `window` in development.

That means you can just type `state.notices.lastTimeShown` in a console rather than installing any of the redux debug tools.

**Comment by travisw** [comment]
> This is looking good to me and works well.
> 
> The notice doesn't seem to appear for very long (maybe only a few seconds?) but that's probably fine.

Thanks @markryall!
It should show for 10 seconds. It could probably show longer... I'm getting some more feedback on that. It can always be adjusted later too.

**Pull request merged**